Monika After Story/game/styles.rpy

```diff
@@ -388,3 +390,146 @@ init -10 python in mas_ui:
             if data["return_value"] == data["true_value"]:
                 return selected_prompt
         return default_prompt
+
+    # Methods for twopane_scrollable_menu
+    def _twopane_menu_filter_events(ev, search_query, search_kws, only_pool, only_random, only_unseen, only_seen):
+        """
+        A method to use as the filter for events in the twopane menu
+
+        IN:
+            ev - event object
+            search_query - search query to filter by
+            search_kws - search_query splitted using spaces
+
+        OUT:
+            boolean whether or not the event pass the criteria
+        """
+        ev_prompt = ev.prompt.lower()
+        ev_label = ev.eventlabel.lower()
+        ev_cat_full = " ".join(map(str, ev.category)) if ev.category else ""
+
+        return (
+            ev_prompt != ev_label
+            and ev.unlocked
+            # and (ev.pool or ev.random)
+            and not ev.anyflags(store.EV_FLAG_HFNAS)
+            and ev.checkAffection(store.mas_curr_affection)
+            and (
+                (not only_pool or ev.pool)
+                and (not only_random or ev.random)
+                and (not only_unseen or ev.shown_count == 0)
+                and (not only_seen or ev.shown_count > 0)
+            )
+            and (
+                search_query in ev_prompt
+                or search_query in ev_label
+                or search_query in ev_cat_full
+                or all(
+                    [
+                        search_kw in ev_prompt
+                        or search_kw in ev_label
+                        or search_kw in ev_cat_full
+                        for search_kw in search_kws
+                    ]
```

Comment:
The one thing I don't like about this is that the list has to be completely generated before it is checked, which means we miss out on short-circuiting the loop when a single condition fails.

If this function is changed to look like Event's filtering function, then short circuiting can be done here so we don't have to loop all the way. 

But we can probably get by with how this is currently if the feel isn't too bad, so I'll reevaluate this when I test this out.

Reply:
If you mean make this similar to `_filterEvent`, then I'm not sure how to do so. That `all` check is part of a bigger, inclusive condition. Right now we generate that list only as the last resort when all the previous conditions have failed, so it's not as bad. If we were to iter the list prior to that, we'd have to create it even when we don't need to check it (unless I'm missing something?). Also note that the list is usually small and we can reuse the search results using cache.

Ideally we do want to make this as optimized as we can, since it's being executed inside screens (which are already slow in RenPy), not to mention we're doing it on the fly (as the user types in) so any lag spike would be noticeable.

Reply:
basically something like
```python
if ev_prompt == ev_label:
    return False

if not ev.unlocked:
    return False

if ev.anyflags(store.EV_FLAG_HFNAS):
    return False

if not ev.checkAffection(store.mas_curr_affection):
    return False

if only_pool and not ev.pool:
    return False

if only_random and not ev.random:
    return False

if only_unseen and ev.shown_count > 0:
    return False
    
if only_seen and ev.shown_count == 0:
    return False

# the final conditions

if search_query in ev_prompt:
    return True

if search_query in ev_label:
    return True

if search_query in ev_cat_full:
    return True

# now we can do the loop
for search_kw in search_kws:
    if (
        search_kw not in ev_prompt
        and search_kw not in ev_label 
        and search_kw not in ev_cat_full
    ):
        return False

return True
```

This allows the loop to short-circuit early. Using `all` requires that the loop needs to be completely generated, which is wasteful.

Reply:
there is also an option of making the loop a separate function that quits early, but I prefer individual if statements when a filter check gets really lengthy like here.

Reply:
Ah, I didn't think about returning `False` early like that, I like it.